File: src/main/java/leader/population/Population.java
package leader.population;

import java.util.*;
import java.util.logging.Logger;

import leader.util.Util;

public class Population {
    private static Logger logger = Logger.getLogger(Population.class.getName());
    private static final int PERSON_BYTE_SIZE = 2;

    private List<byte[]> persons;

    public Population() {
        persons = new ArrayList<>();
    }

    public int size() {
        return persons.size();
    }

    public void clear() {
        persons.clear();
    }

    public byte[] addRandomPerson() {
        return addRandomPerson(Util.randInt(PersonAttribute.ELDERLY.getValue() + 1));
    }

    public byte[] addRandomPerson(int age) {
        byte[] person = new byte[PERSON_BYTE_SIZE];
        pack(PersonAttribute.LIVING.getValue(), person, PersonStruct.ALIVE);
        pack(Util.rand() < 0.5 ? PersonAttribute.FEMALE.getValue(): PersonAttribute.MALE.getValue(), person, PersonStruct.GENDER);
        pack(age, person, PersonStruct.AGE);
        logger.info("Created Person " + explodePerson(person));
        persons.add(person);
        return person;
    }

    public List<byte[]> addRandomPeople(int count) {
        List<byte[]> people = new ArrayList<>();
        for (int i = 0; i < count; ++i) {
            people.add(addRandomPerson());
        }
        return people;
    }

    public List<byte[]> addRandomPeople(int count, int age) {
        List<byte[]> people = new ArrayList<>();
        for (int i = 0; i < count; ++i) {
            people.add(addRandomPerson(age));
        }
        return people;
    }

    public PopulationStatistics generateStats() {
        PopulationStatistics stats = new PopulationStatistics();
        for (byte[] person : persons) {
            Map<PersonStruct, PersonAttribute> attrs = decode(person);
            stats.inc(PersonAttribute.MALE, (attrs.get(PersonStruct.GENDER) == PersonAttribute.MALE ? 1.0 : 0.0));
            stats.inc(PersonAttribute.FEMALE, (attrs.get(PersonStruct.GENDER) == PersonAttribute.FEMALE ? 1.0 : 0.0));
            stats.inc(PersonAttribute.CHILD, (attrs.get(PersonStruct.AGE) == PersonAttribute.CHILD ? 1.0 : 0.0));
            stats.inc(PersonAttribute.ADOLESCENT, (attrs.get(PersonStruct.AGE) == PersonAttribute.ADOLESCENT ? 1.0 : 0.0));
            stats.inc(PersonAttribute.ADULT, (attrs.get(PersonStruct.AGE) == PersonAttribute.ADULT ? 1.0 : 0.0));
            stats.inc(PersonAttribute.ELDERLY, (attrs.get(PersonStruct.AGE) == PersonAttribute.ELDERLY ? 1.0 : 0.0));
        }
        return stats;
    }


    public static Map<PersonStruct, PersonAttribute> decode(byte[] person) {
        Map<PersonStruct, PersonAttribute> attrs = new HashMap<>();
        attrs.put(PersonStruct.ALIVE,  unpack(person, PersonStruct.ALIVE));
        attrs.put(PersonStruct.GENDER, unpack(person, PersonStruct.GENDER));
        attrs.put(PersonStruct.AGE,    unpack(person, PersonStruct.AGE));
        return attrs;
    }


    public static PersonAttribute unpack(byte[] person, PersonStruct type) {
//        int mask = 0;
//        for (int i = 0; i < type.getNumBits(); ++i) {
//            mask |= (1 << i);
//        }
//        mask <<= type.getBitOffset();
//        int v = (person & mask) >> (type.getBitOffset() - 1);
//        return PersonAttribute.toPersonAttribute(type, v);


        int bitOffset = type.getBitOffset();
        int numBits = type.getNumBits();

        int posByte = bitOffset >> 3;
        int posBit = bitOffset & 7;


        long value = 0;
        int modifyBits;
        int valByte;
        int leftShift;
        modifyBits = 8 - posBit;
        if (numBits < modifyBits) modifyBits = numBits;
        leftShift = (8 - posBit - modifyBits);
        while (true) {
            valByte = person[posByte] & 0xff;
            if (modifyBits == 8)
                value += valByte;
            else
                value += (valByte & ((1 << modifyBits) - 1) << leftShift) >> leftShift;
            numBits -= modifyBits;
            if (numBits == 0)
                break;
            posByte++;
            modifyBits = 8;
            if (numBits < modifyBits)
            {
                modifyBits = numBits;
                leftShift = (8 - modifyBits);
            }
            value <<= modifyBits;
        }
        return PersonAttribute.toPersonAttribute(type, (int) value);
    }


    public static void pack(int value, byte[] person, PersonStruct type) {
//        int mask = 0;
//        for (int i = 0; i < type.getNumBits(); ++i) {
//            mask |= (1 << i);
//        }
//        mask <<= type.getBitOffset();
//        person |= mask;
//        return person;
        int numBits = type.getNumBits();
        int bitOffset = type.getBitOffset();

        final long valLong = (value & ((1L << numBits) - 1L));
        int posByte = bitOffset >> 3;
        int posBit = bitOffset & 7;
        int valByte;
        int modifyBits;

        long lValue;
        int leftShift;
        modifyBits = 8 - posBit;
        if (numBits < modifyBits)
            modifyBits = numBits;
        leftShift = (8 - posBit - modifyBits);
        while (true){
            valByte = person[posByte];
            if (modifyBits == 8) {
                lValue = valLong << (32 - numBits) >> (24);
                person[posByte] = (byte) lValue;
            }
            else {
                lValue = valLong << (32 - numBits) >> (32 - modifyBits) << leftShift;
                person[posByte] = (byte) ((valByte & ~(((1 << modifyBits) - 1) << leftShift)) | lValue);
            }
            numBits -= modifyBits;
            if (numBits == 0) break;
            posByte++;
            modifyBits = 8;
            if (numBits < modifyBits) {
                modifyBits = numBits;
                leftShift = (8 - modifyBits);
            }
        }
    }

    // Advance this person's age.  If the person died, return False.  Return True otherwise.
    public static void advanceAge(byte[] person) {
        PersonAttribute age = unpack(person, PersonStruct.AGE);
        if (age != PersonAttribute.ELDERLY)
            pack(age.getValue() + 1, person, PersonStruct.AGE);
    }

    public void buryTheDead(){
        Iterator<byte[]> iterator = persons.iterator();
        while (iterator.hasNext()){
            byte[] person = iterator.next();
            if (unpack(person, PersonStruct.ALIVE) == PersonAttribute.DEAD) {
                iterator.remove();
            }
        }
    }

    public List<byte[]> getPersons() {
        return persons;
    }

    public String explodePerson(byte[] person){
        StringBuilder sb = new StringBuilder();
        for (byte b: person) {
            byte mask = (byte) (1 << (Byte.SIZE - 1));
            for (int i = 0; i < Byte.SIZE; ++i) {
                sb.append((b & mask) > 0 ? "1" : "0");
                mask >>= 1;
            }
        }
        return sb.toString();
    }
}

File: python/population.py


class Population( object ):
    # Statistics
    MALE   = "Male"
    FEMALE = "Female"
    CHILD  = "Child"
    ADOLESCENT = "Adolescent"
    ADULT = "Adult"
    ELDERLY = "Elderly"
    
    # Person attributes
    GENDER = "Gender"
    AGE = "Age"
    
    # Bit unpacking parameters (byte, bitoffset, numbits)
    GENDER_DECODE = (0, 0, 1)  # 0: female, 1: male
    AGE_DECODE    = (0, 2, 3)  # Decade of life: 0: child, 1: adolescent, 2-5: adult, 6-7: elderly
    
    def __init__( self ):
        pass
        
    def generate_stats( population ):
        stats = {
            self.MALE: 0,
            self.FEMALE: 0,
            self.CHILD: 0,
            self.ADOLESCENT: 0,
            self.ADULT: 0,
            self.ELDERLY: 0,
        }
        for person in population:
            attrs = self.decode( person )
            stats[ self.MALE       ] += 1 if attrs[ self.GENDER ] == self.MALE else 0
            stats[ self.FEMALE     ] += 1 if attrs[ self.GENDER ] == self.FEMALE else 0
            stats[ self.CHILD      ] += 1 if attrs[ self.AGE    ] == self.CHILD else 0
            stats[ self.ADOLESCENT ] += 1 if attrs[ self.AGE    ] == self.ADOLESCENT else 0
            stats[ self.ADULT      ] += 1 if attrs[ self.AGE    ] == self.ADULT else 0
            stats[ self.ELDERLY    ] += 1 if attrs[ self.AGE    ] == self.ELDERLY else 0
    generate_stats = staticmethod( generate_stats )
            
            
    def decode( person ):
        ubytes = struct.unpack( "<BBBB", person )
        attrs = {
            self.GENDER: self.MALE if self.unpack( ubytes, GENDER_DECODE ) > 0 else self.FEMALE,
            self.AGE: self.unpack_age( ubytes, AGE_DECODE ),
        }
        return attrs
    decode = staticmethod( decode )
    
    
    def encode( person, attr, value ):
        params = None
        if attr == self.GENDER:
            params = self.GENDER_DECODE
        elif attr == self.AGE:
            params = self.AGE_DECODE
        if params is None:
            print "Unable to set %s value: cannot find decode parameters" % attr
            return person
        return self.pack( value, person, params )
    encode = staticmethod( encode )
    
        
    def unpack( ubytes, decode_param ):
        mask = 0
        for i in xrange( decode_param[ 2 ] ):
            mask |= (1 << i)
        mask <<= decode_param[ 1 ]
        v = (ubytes[ decode_param[ 0 ] ] & mask) >> (decode_param[ 2 ] - 1)
        return v
    unpack = staticmethod( unpack )
    
        
    def unpack_age( ubytes, decode_param ):
        v = self.unpack( ubytes, decode_param )
        if v == 0:
            return self.CHILD
        elif v == 1:
            return self.ADOLESCENT
        elif 2 >= v >= 5:
            return self.ADULT
        return self.ELDERLY
    unpack_age = staticmethod( unpack_age )
    
        
    def pack( value, person, decode_param ):
        ubytes = struct.unpack( "<BBBB", person )
        mask = 0
        for i in xrange( decode_param[ 2 ] ):
            mask |= (1 << i)
        mast <<= decode_param[ 1 ]
        ubytes[ decode_param[ 0 ] ] |= mask
        person = struct.pack( "<BBBB", *ubytes )
        return person
    pack = staticmethod( pack )
    
    
    def advance_age( person ):
        """ Advance this person's age.  If the person died, return False.  Return True otherwise. """
        age = self.unpack_age( person )
        if age == self.ELDERLY:
            return person  # Elderly people stay elderly!
        return encode( person, AGE, age + 1 )
    